### PR TITLE
[New Rule] Building Block - AWS Lambda Function Created or Updated

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/ia-trade-team/issues/358

## Summary
Identifies when an AWS Lambda function is created or updated. AWS Lambda lets you run code without provisioning or
managing servers. Adversaries can create or update Lambda functions to execute malicious code, exfiltrate data, or
escalate privileges. 

This is a building block rule due to the expected frequency at which Lambda functions could be created or updated. This is expected by dev teams in AWS and thus would be too noisy as a generic rule. Instead it serves better as a signal. Additionally, there are no indications from the telemetry that this could be malicious to eliminate potential false positives and there is no code visibility when updating that we could query on.

Screenshot 2024-05-28 at 11 35 04 AM

<details><summary>Example Data</summary>

```
{
  "_index": ".ds-logs-aws.cloudtrail-default-2024.04.09-000001",
  "_id": "REDACTED",
  "_version": 1,
  "_score": 0,
  "_source": {
    "agent": {
      "name": "REDACTED",
      "id": "REDACTED",
      "type": "filebeat",
      "ephemeral_id": "REDACTED",
      "version": "8.13.2"
    },
    "log": {
      "file": {
        "path": "REDACTED"
      },
      "offset": 1
    },
    "elastic_agent": {
      "id": "REDACTED",
      "version": "8.13.2",
      "snapshot": false
    },
    "source": {
      "geo": {
        "region_iso_code": "US-OH",
        "continent_name": "North America",
        "city_name": "REDACTED",
        "country_iso_code": "US",
        "country_name": "United States",
        "region_name": "Ohio",
        "location": {
          "lon": -81.3798,
          "lat": 40.7961
        }
      },
      "as": {
        "number": 10796,
        "organization": {
          "name": "REDACTED"
        }
      },
      "address": "REDACTED",
      "ip": "REDACTED"
    },
    "tags": [
      "preserve_original_event",
      "forwarded",
      "aws-cloudtrail"
    ],
    "cloud": {
      "region": "us-east-1",
      "account": {
        "id": "REDACTED"
      }
    },
    "input": {
      "type": "aws-s3"
    },
    "@timestamp": "2024-04-30T22:27:42.000Z",
    "ecs": {
      "version": "8.0.0"
    },
    "related": {
      "user": [
        "REDACTED"
      ]
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "aws.cloudtrail"
    },
    "tls": {
      "cipher": "TLS_AES_128_GCM_SHA256",
      "client": {
        "server_name": "REDACTED"
      },
      "version": "1.3",
      "version_protocol": "tls"
    },
    "event": {
      "agent_id_status": "verified",
      "ingested": "2024-04-30T22:33:12Z",
      "original": "REDACTED",
      "provider": "lambda.amazonaws.com",
      "created": "2024-04-30T22:33:09.701Z",
      "kind": "event",
      "action": "CreateFunction20150331",
      "id": "REDACTED",
      "type": [
        "info"
      ],
      "dataset": "aws.cloudtrail",
      "outcome": "success"
    },
    "aws": {
      "s3": {
        "bucket": {
          "name": "REDACTED",
          "arn": "REDACTED"
        },
        "object": {
          "key": "REDACTED"
        }
      },
      "cloudtrail": {
        "event_version": "1.08",
        "flattened": "REDACTED",
        "event_type": "AwsApiCall",
        "read_only": false,
        "user_identity": "REDACTED",
        "recipient_account_id": "REDACTED",
        "management_event": true
      }
    },
    "user": {
      "name": "REDACTED",
      "id": "REDACTED"
    },
    "user_agent": {
      "original": "REDACTED",
      "name": "aws-sdk-go",
      "device": {
        "name": "Other"
      },
      "version": "1.44.157"
    }
  },
  "fields": {
    "aws.cloudtrail.request_parameters.text": "REDACTED",
    "elastic_agent.version": "8.13.2",
    "tls.version_protocol": "tls",
    "user_agent.original.text": "REDACTED",
    "agent.name.text": "REDACTED",
    "source.geo.region_name": "Ohio",
    "source.ip": "REDACTED",
    "agent.name": "REDACTED",
    "user_agent.version": "1.44.157",
    "event.agent_id_status": "verified",
    "source.geo.region_iso_code": "US-OH",
    "aws.cloudtrail.management_event": "true",
    "event.kind": "event",
    "aws.cloudtrail.user_identity.arn": "REDACTED",
    "event.outcome": "success",
    "source.geo.city_name": "REDACTED",
    "tls.version": "1.3",
    "user_agent.original": "REDACTED",
    "event.original": "REDACTED",
    "cloud.region": "us-east-1",
    "user.id": "REDACTED",
    "input.type": "aws-s3",
    "log.offset": 1,
    "user_agent.name": "aws-sdk-go",
    "data_stream.type": "logs",
    "related.user": "REDACTED",
    "tags": [
      "preserve_original_event",
      "forwarded",
      "aws-cloudtrail"
    ],
    "event.provider": "lambda.amazonaws.com",
    "agent.id": "REDACTED",
    "ecs.version": "8.0.0",
    "event.created": "2024-04-30T22:33:09.701Z",
    "aws.cloudtrail.event_version": "1.08",
    "agent.version": "8.13.2",
    "source.as.number": 10796,
    "aws.cloudtrail.read_only": false,
    "aws.cloudtrail.event_category": "Management",
    "aws.cloudtrail.user_identity.type": "IAMUser",
    "aws.s3.bucket.arn": "REDACTED",
    "aws.cloudtrail.recipient_account_id": "REDACTED",
    "aws.cloudtrail.request_id": "REDACTED",
    "tls.cipher": "TLS_AES_128_GCM_SHA256",
    "user.name": "REDACTED",
    "source.geo.location": {
      "coordinates": [
        -81.3798,
        40.7961
      ],
      "type": "Point"
    },
    "source.address": "REDACTED",
    "aws.cloudtrail.flattened.request_parameters": "REDACTED",
    "agent.type": "filebeat",
    "event.module": "aws",
    "source.geo.country_iso_code": "US",
    "aws.cloudtrail.response_elements.text": "REDACTED",
    "elastic_agent.snapshot": false,
    "aws.cloudtrail.event_type": "AwsApiCall",
    "aws.s3.bucket.name": "REDACTED",
    "source.as.organization.name.text": "REDACTED",
    "elastic_agent.id": "REDACTED",
    "data_stream.namespace": "default",
    "source.as.organization.name": "REDACTED",
    "source.geo.continent_name": "North America",
    "tls.client.server_name": "REDACTED",
    "event.action": "CreateFunction20150331",
    "event.ingested": "2024-04-30T22:33:12.000Z",
    "@timestamp": "2024-04-30T22:27:42.000Z",
    "cloud.account.id": "REDACTED",
    "aws.cloudtrail.user_identity.access_key_id": "REDACTED",
    "data_stream.dataset": "aws.cloudtrail",
    "event.type": "info",
    "log.file.path": "REDACTED",
    "agent.ephemeral_id": "REDACTED",
    "aws.cloudtrail.request_parameters": "REDACTED",
    "event.id": "REDACTED",
    "source.geo.country_name": "United States",
    "user_agent.device.name": "Other",
    "aws.s3.object.key": "REDACTED",
    "event.dataset": "aws.cloudtrail",
    "user.name.text": "REDACTED"
  }
}

```

</details>

### CODE REVIEW

File: rules_building_block/execution_aws_lambda_function_updated.toml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+bypass_bbr_timing = true
+
+[rule]
+author = ["Elastic"]
+building_block_type = "default"
+description = """
+Identifies when an AWS Lambda function is created or updated. AWS Lambda lets you run code without provisioning or
+managing servers. Adversaries can create or update Lambda functions to execute malicious code, exfiltrate data, or
+escalate privileges. This is a [building block
+rule](https://www.elastic.co/guide/en/security/current/building-block-rule.html) that does not generate alerts, but
```
Comment (Aegrah) [inline rules_building_block/execution_aws_lambda_function_updated.toml:17]: Technically BBRs do generate alerts right?

File: rules_building_block/execution_aws_lambda_function_updated.toml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    "Use Case: Asset Visibility",
+    "Tactic: Execution",
+]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.dataset: "aws.cloudtrail"
+    and event.provider: lambda.amazonaws.com and event.outcome: success
+    and event.action: (CreateFunction* or UpdateFunctionCode*)
```
Comment (Aegrah) [inline rules_building_block/execution_aws_lambda_function_updated.toml:55]: You might want to try to keep the formatting consistent across all of the AWS rules.

**Review by Aegrah — APPROVED** [review]
LGTM!

File: rules_building_block/execution_aws_lambda_function_updated.toml (hunk 1 above)
Comment (terrancedejesus) [reply rules_building_block/execution_aws_lambda_function_updated.toml:17]: They generate events in the signal index if I recall.

File: rules_building_block/execution_aws_lambda_function_updated.toml (hunk 2 above)
Comment (terrancedejesus) [inline rules_building_block/execution_aws_lambda_function_updated.toml:55]:
```suggestion
event.dataset: "aws.cloudtrail"
    and event.provider: "lambda.amazonaws.com"
    and event.outcome: "success"
    and event.action: (CreateFunction* or UpdateFunctionCode*)
```

Comment (terrancedejesus) [reply rules_building_block/execution_aws_lambda_function_updated.toml]: Updated it to be consistent. Thanks Ruben!

**Review by imays11 — APPROVED** [review]

**Pull request merged**